test/unit/matrix/test_matrix.cpp

```diff
@@ -1275,6 +1280,29 @@ TEST(MatrixDestructorFutures, NonConstAfterRead) {
   last_task.get();
 }
 
+TEST(MatrixDestructorFutures, NonConstAfterReadWithSenderAdaptors) {
+  namespace ex = pika::execution::experimental;
+  pika::future<void> last_task;
+
+  std::atomic<bool> is_exited_from_scope{false};
+  {
+    auto matrix = createMatrix<TypeParam>();
+
+    auto shared_future = matrix.read_sender(LocalTileIndex(0, 0));
+    last_task = dlaf::internal::transform(
+                    dlaf::internal::Policy<dlaf::Backend::MC>(),
+                    [&is_exited_from_scope](auto&&) {
+                      try_waiting_guard(is_exited_from_scope);
+                      EXPECT_TRUE(is_exited_from_scope);
+                    },
```

Comment:
I'm not quite sure what you're proposing. Could maybe roughly sketch the code that you're imagining?

Reply:
I was thinking about creating a function that accepts a reference to the guard, so that we can reuse it all over our tests.

Just a draft for sharing the idea, I'm pretty sure compiler won't be on my side.
```cpp
void waitAndSet(auto&&, auto& guard) {
  try_waiting_guard(guard);
  EXPECT_TRUE(guard);
}

last_task =
  dlaf::internal::whenAllLift(
    shared_future,
    std::reference_wrapper(is_exited_from_scope))
  | dlaf::internal::transform(dlaf::internal::Policy<dlaf::Backend::MC>(), waitAndSet)
  | make_future();
```

Reply:
What do you think about
```
struct Waiter {
  std::atomic<bool>& is_exited_from_scope;
  template <typename T>
  void operator()(T&& t) {
    try_waiting_guard(is_exited_from_scope);
    EXPECT_TRUE(is_exited_from_scope);
  }
};

// ...

last task = shared_future |
  transform(Policy<Backend::MC>(), Waiter{is_exited_from_scope}) |
  make_future();
```
? Also not compiled... but `Waiter` would also work as a drop-in replacement in the `future::then` versions of the tests as well. They could of course also be changed to use `when_all` with your version.

I'm not set on either of them so I'll change this to whatever you prefer.

Reply:
Yeah, it was just an idea, and not a real request.

I think you should go for whatever it makes more sense for the code that will use it. If `then` is a use-case, I'm completely fine with the callable that does not require the guard as argument. Indeed, my point was about having re-usable code.

Reply:
I think it's a good idea. Somewhat unrelated, but small enough that I'll happily do the change here. I'll try to push something in a bit.

Reply:
I moved this out into a helper function/struct. Please have a look to see if you agree with the implementation (and the name!).

Reply:
super nice! about the name it can be left like this. thanks